### PR TITLE
Allow re-creation of the shared context as a core context

### PR DESCRIPTION
**Pull request by binary1248**
Allow re-creation of the shared context as a core context if the user indicates they want a core profile context. Sharing of compatibility and core profile contexts is not possible on macOS which is why we need to have a way to re-create the shared context as a core context if required in this case.

The following code breaks on macOS without this change:
```cpp
#include <SFML/Window.hpp>

int main()
{
    sf::Window window(sf::VideoMode(100, 100), "Test", sf::Style::Default, sf::ContextSettings(0, 0, 0, 3, 2, sf::ContextSettings::Core));

    while (window.isOpen())
    {
        sf::Event event;
        while (window.pollEvent(event))
        {
            if (event.type == sf::Event::Closed)
                window.close();
        }

        window.display();
    }
}
```

### CODE REVIEW

File: src/SFML/Window/GlContext.cpp [hunk 1]
```diff
@@ -212,6 +212,55 @@ namespace
...
+        if (glGetError() == GL_INVALID_ENUM)
+        {
+            // Try to load the < 3.0 way
+            const char* extensionString = reinterpret_cast<const char*>(glGetString(GL_EXTENSIONS));
+
+            do
+            {
+                const char* extension = extensionString;
+
+                while(*extensionString && (*extensionString != ' '))
```
Comment (mantognini) [inline src/SFML/Window/GlContext.cpp]: missing space: `while_(`

File: src/SFML/Window/GlContext.cpp [hunk 2]
```diff
@@ -446,6 +471,25 @@ GlContext* GlContext::create(const ContextSettings& settings, unsigned int width
 
     Lock lock(mutex);
 
+    // If resourceCount is 1 we know that we are inside sf::Context or sf::Window
```
Comment (mantognini) [inline src/SFML/Window/GlContext.cpp]: DRY code: re-factor this code and the identical one from above into a function. Maybe a template function to accommodate for the width/height vs owner/bpp difference? If the logic is the same (except how `context` is constructed), I think it makes sense.

File: src/SFML/Window/Win32/WglContext.cpp [hunk 3]
```diff
@@ -137,17 +149,23 @@ m_ownsWindow   (false)
...
-    // Make sure that extensions are initialized if this is not the shared context
-    // The shared context is the context used to initialize the extensions
-    if (shared && shared->m_deviceContext)
-        ensureExtensionsInit(shared->m_deviceContext);
-
     // Create the rendering surface (window or pbuffer if supported)
     createSurface(shared, width, height, VideoMode::getDesktopMode().bitsPerPixel);
 
     // Create the context
     if (m_deviceContext)
```
Comment (mantognini) [inline src/SFML/Window/Win32/WglContext.cpp]: If possible, I would also DRY the ctors.

File: src/SFML/Window/GlContext.cpp (hunk 2 above)
Comment (binary1248) [reply src/SFML/Window/GlContext.cpp]: I initially tried to do this, but `GlContext::initialize()` is private.

File: src/SFML/Window/Win32/WglContext.cpp (hunk 3 above)
Comment (binary1248) [reply src/SFML/Window/Win32/WglContext.cpp]: Delegating constructors are coming in C++11. 😉

Comment (mantognini) [reply src/SFML/Window/Win32/WglContext.cpp]: Good! Maybe add a `\todo` so we actually clean up the code at some point?

**Comment by binary1248** [comment]
There... should be a bit better now. The amount of code duplication can probably be significantly reduced in C++11. I wouldn't invest too much time in trying to minimize the lines of code right now.

**Pull request merged**